fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+
+        This is intended to be used with an embedding layer with shared weights.
+
+    Args:
+        proj_weight (nn.Parameter):
+            The shared weight.
+        tile_factor (int):
+            Unused. It is here to make kernel init easier with MEVO.
+        log (bool):
+            If True, use log_softmax instead of softmax.
+
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__()
+        out_dim, in_dim = proj_weight.shape
+        assert "cuda" in str(proj_weight.device), "weight should be on GPU"
+        self.fc = nn.Linear(in_dim, out_dim, bias=False).to("cuda")
+        assert proj_weight.dtype in [torch.float16, torch.float32]
+        if proj_weight.dtype == torch.float16:
+            self.fc = self.fc.half()
+        self.fc.weight = proj_weight
+        assert self.fc.weight.dtype in [torch.float16, torch.float32], self.fc.weight.dtype
+        self.fp16 = self.fc.weight.dtype == torch.float16
+        self.log = log
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
+        """ Forward function that computes softmax output with the input and target."""
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        if self.fp16:
+            assert input.dtype == torch.float16
+        x = self.fc(input)
+        # Note that we do softmax in FP32, which is important for numerical stability.
+        if self.log:
+            x = F.log_softmax(x, dim=-1, dtype=torch.float32)
+        else:
+            x = F.softmax(x, dim=-1, dtype=torch.float32)
+        assert x.dtype == torch.float32
+        return x
+
+
+class BaselineSoftmaxNllLoss(BaselineSoftmax):
+    """ Baseline that does an output projection, a softmax & a NLL loss (cross-entropy).
+
+        See BaselineSoftmax above. Constructor is the same. Only difference is in the
+        forward function.
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
+        """Forward that directly compute the loss."""
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        x = super().forward(input, target)
+        x = F.nll_loss(x, target, reduction="sum")
+        return x
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
+
+    @staticmethod
+    def get_max(i: torch.Tensor, w: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _m = torch.matmul(i, w.T)
+        if fp:
```

Comment:
what does fp stand for?

Reply:
fp is full precision, sharing the same acronym with floating point. :-)

I will make those longer names.